dht_bootstrap.go

```diff
@@ -41,52 +43,47 @@ func init() {
 	}
 }
 
-// BootstrapConfig runs cfg.Queries bootstrap queries every cfg.BucketPeriod.
-func (dht *IpfsDHT) Bootstrap(ctx context.Context) error {
-	triggerBootstrapFnc := func() {
-		logger.Infof("triggerBootstrapFnc: RT only has %d peers which is less than the min threshold of %d, triggering self & bucket bootstrap",
-			dht.routingTable.Size(), minRTBootstrapThreshold)
-
-		if err := dht.selfWalk(ctx); err != nil {
-			logger.Warningf("triggerBootstrapFnc: self walk: error: %s", err)
-		}
+// Bootstrap  i
+func (dht *IpfsDHT) startBootstrapping() error {
+	// scan the RT table periodically & do a random walk on k-buckets that haven't been queried since the given bucket period
+	dht.proc.Go(func(proc process.Process) {
+		ctx := processctx.OnClosingContext(proc)
+		scanInterval := time.NewTicker(dht.bootstrapCfg.RoutingTableScanInterval)
+		defer scanInterval.Stop()
 
-		if err := dht.bootstrapBuckets(ctx); err != nil {
-			logger.Warningf("triggerBootstrapFnc: bootstrap buckets: error bootstrapping: %s", err)
-		}
-	}
+		var (
+			lastSelfWalk time.Time
+			walkSelf     = true
+		)
 
-	// we should query for self periodically so we can discover closer peers
-	go func() {
 		for {
-			err := dht.selfWalk(ctx)
-			if err != nil {
-				logger.Warningf("self walk: error: %s", err)
-			}
-			select {
-			case <-time.After(dht.bootstrapCfg.SelfQueryInterval):
-			case <-ctx.Done():
-				return
+			if walkSelf {
```

Comment:
I combined these into a single loop.

1. We always want to finish `selfWalk` _before_ checking the buckets.
2. It doesn't make sense to walk self if we're not going to immediately fill in the empty buckets.